core/test/solver/chebyshev.cpp

```diff
@@ -0,0 +1,375 @@
+    using Solver = typename TestFixture::Solver;
+    this->solver->clear();
+
+    ASSERT_EQ(this->solver->get_size(), gko::dim<2>(0, 0));
+    auto solver_mtx =
+        static_cast<Solver*>(this->solver.get())->get_system_matrix();
+    ASSERT_EQ(solver_mtx, nullptr);
+}
+
+
+TYPED_TEST(Chebyshev, DefaultApplyUsesInitialGuess)
+{
+    ASSERT_TRUE(this->solver->apply_uses_initial_guess());
+}
+
+
+TYPED_TEST(Chebyshev, CanSetEigenRegion)
+{
+    using Solver = typename TestFixture::Solver;
+    using value_type = typename TestFixture::value_type;
+    std::shared_ptr<Solver> chebyshev_solver =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(3u).on(this->exec))
+            .with_foci(value_type{0.2}, value_type{1.2})
+            .on(this->exec)
+            ->generate(this->mtx);
+
+    ASSERT_EQ(chebyshev_solver->get_parameters().foci,
+              std::make_pair(value_type{0.2}, value_type{1.2}));
+}
+
+
+TYPED_TEST(Chebyshev, CanSetInnerSolverInFactory)
+{
+    using Solver = typename TestFixture::Solver;
+    using value_type = typename TestFixture::value_type;
+    auto chebyshev_factory =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(3u).on(this->exec),
+                gko::stop::ResidualNorm<value_type>::build()
+                    .with_reduction_factor(r<value_type>::value)
+                    .on(this->exec))
+            .with_solver(
+                Solver::build()
+                    .with_criteria(
+                        gko::stop::Iteration::build().with_max_iters(3u).on(
+                            this->exec))
+                    .on(this->exec))
+            .on(this->exec);
+    auto solver = chebyshev_factory->generate(this->mtx);
+    auto inner_solver = dynamic_cast<const Solver*>(
+        static_cast<Solver*>(solver.get())->get_solver().get());
+
+    ASSERT_NE(inner_solver, nullptr);
+    ASSERT_EQ(inner_solver->get_size(), gko::dim<2>(3, 3));
+    ASSERT_EQ(inner_solver->get_system_matrix(), this->mtx);
+}
+
+
+TYPED_TEST(Chebyshev, CanSetGeneratedInnerSolverInFactory)
+{
+    using Solver = typename TestFixture::Solver;
+    std::shared_ptr<Solver> chebyshev_solver =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(3u).on(this->exec))
+            .on(this->exec)
+            ->generate(this->mtx);
+
+    auto chebyshev_factory =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(3u).on(this->exec))
+            .with_generated_solver(chebyshev_solver)
+            .on(this->exec);
+    auto solver = chebyshev_factory->generate(this->mtx);
+    auto inner_solver = solver->get_solver();
+
+    ASSERT_NE(inner_solver.get(), nullptr);
+    ASSERT_EQ(inner_solver.get(), chebyshev_solver.get());
+}
+
+
+TYPED_TEST(Chebyshev, CanSetCriteriaAgain)
```

Comment:
```suggestion
TYPED_TEST(Chebyshev, CanSetCriteria)
```

Reply:
I prefer staying the `CanSetCriteriaAgain` because it set criteria first from factory and then reset it by member function